### DISCUSSION TITLE
player.matchStats does not work properly with overtime matches (Faceit)

### DISCUSSION THREAD

**Discussion by thecatontheflat** [saul/demofile#259] [Questions] [answered]
**Describe the bug**
I suspect this issue is related to this one https://github.com/saul/demofile/issues/60

I am using `player.matchStats` to get a brief summary of the player performance. However I noticed that it does not work well with games with overtimes (Faceit). While observing the values coming to the `m_iMatchStats_*` tables, specifically `m_iMatchStats_Kills`, I noticed that during the overtime, first array elements get replaced after round 30 (e.g. round 35 is located in player.matchDetails[4]).

It wasn't a problem, since I simply implemented a logic to "reset" the index of the current round stats, and aggregated them in a separate array during each `round_end` event. However, in this specific demo, there is a discrepancy between player kills based on the game events and based on the `m_iMatchStats_Kills` values. Round 34 and player steam64id `76561198932901854` shows 3 kills in their `m_iMatchStats_Kills` table, however they only actually make 2 kills.

I suspect this is just how server sends the data and there is not much I can do except tracking each stat manually?

**Code to reproduce**
```
demoFile.gameEvents.on('cs_win_panel_match', e => {
const matchStats = demoFile.players[0].matchStats;
  console.log(demoFile.gameRules.roundNumber) // 35
  console.log(matchStats.length); // 30
});
```

**Expected behaviour**
`player.matchStats` size corresponds to the number of rounds played in the match

**Affected demos**
https://demos-us-central1.faceit-cdn.net/csgo/d938ad36-71cf-4773-a9a9-4441e732ab1a.dem.gz

**Comment by saul** [comment]
Unfortunately this is just how it is networked. As for the first issue (index wrap around), I maybe wouldn't rely on the returned array of `matchStats`, and instead listen for when any of the props change. When any of the props change, save the state of the most recent `matchStats`. I imagine the way it's been implemented on Valve's side is that it's just a 30 element array (5 bits to index). You could possibly lobby them to increment this to 6/7 bits (64/128 elements).

> Round 34 and player steam64id 76561198932901854 shows 3 kills in their m_iMatchStats_Kills table, however they only actually make 2 kills.

What leads you to believe that they made only 2 kills? Again this prop is networked, so I'd trust this prop over counting `player_death` events.

**Comment by thecatontheflat** [comment]
> What leads you to believe that they made only 2 kills? Again this prop is networked, so I'd trust this prop over counting player_death events.

I thought so too, but when I noticed this discrepancy I actually watched the demo and indeed they made only 2 kills in that round (which also corresponds to the `player_death` events). Here is the sample code (ignore the warmup). Look for round 34:

Round: 34
m_iMatchStats_Kills: 3
Actual kills: 2



```
const samplePlayerId = '76561198932901854';
let actualKillsInTheRound = 0;
demoFile.gameEvents.on('player_death', e => {
  const attacker = demoFile.entities.getByUserId(e.attacker);

  if (attacker.steam64Id === samplePlayerId) {
    actualKillsInTheRound++;
  }
});


demoFile.gameEvents.on('round_end', e => {
  const roundNumber = demoFile.gameRules.roundNumber;
  demoFile.players.forEach(player => {
    if (player.steam64Id === samplePlayerId) {

      const matchStats = player.matchStats;
      const roundResetNumber = roundNumber <= 30 ? roundNumber : roundNumber % 30;
      const roundIndex = roundResetNumber - 1;
      const roundStats = matchStats[roundIndex];

      console.log('Round:', roundNumber);
      console.log('m_iMatchStats_Kills:', roundStats.kills);
      console.log('Actual kills:', actualKillsInTheRound);
      console.log('')

    }
  });

  actualKillsInTheRound = 0;
});
```

Interestingly enough, the game client displays the match stats correctly. It seems they don't really use `m_iMatchStats_*` (or at least not according to my expectations ha)

**Comment by saul** [comment]
Haven't had chance to run this myself yet, but remember players can still be killed after `round_end` - you should be rolling over the match stats at `round_start` instead. Could this be the cause of your problem?

Round end is fired when the round is won (e.g. bomb defuse), but you still have a few seconds of playtime until the next round starts.

**Comment by thecatontheflat** [comment]
Thank you for the suggestion! I will check it and will let you know about the findings. I assumed however that `m_iMatchStats_*` should hold information about the kills as it appears on the scoreboard (meaning only "scored" kills will be present there)

**Comment by thecatontheflat** [comment, marked as answer]
That was it! I replaced the code with the following code where the `prevRoundStats` contain the right number:

```
demoFile.gameEvents.on('round_end', e => {
  const roundNumber = demoFile.gameRules.roundNumber;
  demoFile.players.forEach(player => {
    if (player.steam64Id === samplePlayerId) {

      const matchStats = player.matchStats;
      const roundResetNumber = roundNumber <= 30 ? roundNumber : roundNumber % 30;
      const roundIndex = roundResetNumber - 1;
      const roundStats = matchStats[roundIndex];
      const prevRoundStats = matchStats[roundIndex - 1];

      console.log('Round:', roundNumber);
      console.log('m_iMatchStats_Kills:', roundStats.kills);
      if (prevRoundStats) console.log('previous round m_iMatchStats_Kills:', prevRoundStats.kills);
      console.log('Actual kills:', actualKillsInTheRound);
      console.log('')
    }
  });
});
```

This works for every round except the last one, but I guess it's impossible to have a discrepancy there since the game freezes immediately.

Thanks for the tip!

**Comment by saul** [comment]
Glad that helped!
